File: velolabs_repos/lattis_android/app/src/main/java/com/lattis/ellipse/domain/interactor/lock/observe/ObserveHardwareStateUseCase.java
package com.lattis.ellipse.domain.interactor.lock.observe;

import com.lattis.ellipse.domain.executor.PostExecutionThread;
import com.lattis.ellipse.domain.executor.ThreadExecutor;
import com.lattis.ellipse.domain.interactor.lock.base.BaseLockUseCase;
import com.lattis.ellipse.domain.model.Lock;
import com.lattis.ellipse.domain.repository.BluetoothRepository;
import com.lattis.ellipse.domain.repository.LockRepository;

import javax.inject.Inject;

import io.reactivex.Observable;

public class ObserveHardwareStateUseCase extends BaseLockUseCase<Lock.Hardware.State> {

    @Inject
    ObserveHardwareStateUseCase(ThreadExecutor threadExecutor,
                                PostExecutionThread postExecutionThread,
                                BluetoothRepository bluetoothRepository,
                                LockRepository lockRepository) {
        super(threadExecutor, postExecutionThread,bluetoothRepository,lockRepository);
    }

    public ObserveHardwareStateUseCase forLock(Lock lock) {
        setScannedLock(lock);
        return this;
    }

    @Override
    protected Observable<Lock.Hardware.State> buildUseCaseObservable() {
        return getHardwareStateObservable();
    }
}

File: velolabs_repos/lattis_android_v2/domain/src/main/java/com/lattis/domain/usecase/lock/observe/ObserveHardwareStateUseCase.kt
package com.lattis.domain.usecase.lock.observe

import com.lattis.domain.executor.PostExecutionThread
import com.lattis.domain.executor.ThreadExecutor
import com.lattis.domain.repository.BluetoothRepository
import com.lattis.domain.repository.LockRepository
import com.lattis.domain.usecase.lock.base.BaseLockUseCase
import com.lattis.domain.models.Lock
import com.lattis.domain.repository.SaSOrPSLockRepository
import io.reactivex.rxjava3.core.Observable
import javax.inject.Inject

class ObserveHardwareStateUseCase @Inject internal constructor(
    threadExecutor: ThreadExecutor,
    postExecutionThread: PostExecutionThread,
    bluetoothRepository: BluetoothRepository,
    lockRepository: LockRepository,
    saSOrPSLockRepository: SaSOrPSLockRepository
) : BaseLockUseCase<Lock.Hardware.State>(
    threadExecutor,
    postExecutionThread,
    bluetoothRepository,
    lockRepository,
    saSOrPSLockRepository
) {
    fun withLockVendor(vendor:LockVendor): ObserveHardwareStateUseCase {
        this.lockVendor = vendor
        return this
    }

    fun forLock(lock: Lock): ObserveHardwareStateUseCase {
        setScannedLock(lock)
        return this
    }

    override fun buildUseCaseObservable(): Observable<Lock.Hardware.State> {
        return hardwareStateObservable(lockVendor)
    }
}